luanti-org/luanti
# Network encryption

## Introduction

Implements a network encryption system for reliable packets. ECDHE with x25519 is used to establish a shared secret, which is then used to derive multiple encryption keys that are used to encrypt packet data using ChaCha20-poly1305 with a 16 byte auth tag. A digest of the initial x25519 secret is included in the Secure Remote Password state that is hashed to generate the evidence messages proving knowledge of the password, thereby preventing active eavesdropping by anyone that doesn't know the user's password.

A more detailed description of the cryptosystem is included in `network/encryption.h`.

The related issues would be #10206 and #14903.

Related pull requests:  #14129 and #14196

## Why not certificates/PKI

Anyone should be able to start a Minetest server without having to go through the process of obtaining a CA signed certificate nor would anyone likely want to run a CA for Minetest. Using the existing HTTPS ACME infrastructure is possible but would require a server operator also maintains a web server and configures it correctly, and this would only offer security even connecting to a server using the domain name it's hosted at not just the IP address.

If this pull request is accepted, a future extension of the protocol could go down that route and perhaps should. 

## Why this solution and not dtls, xyz, abc, etc

This solution is built using very similar cryptographic primitives as those protocols but it reuses as much of the existing network implementation as possible meaning two networking backends don't need to be maintained for backwards compatibility. 

Additionally TLS 1.3 removed support for the `ECDH_anon_*` ciphersuits and does not support a SRP or PAKE extension like TLS 1.2 did. There is likely a greater risk of an implementation oversight breaking the cryptosystem doing it like this, but I think that's worthwhile for a simpler implementation.

## What dependencies does this add?

HACL* is partially vendor-ed in, tried to cut down the number of files included as much as possible without modifying any of them.



## Why only reliable packets?

Encrypting unreliable packets would require giving them something like a sequence number so that a IV can be generated. This can be done if there is consensus that it's a good idea, it seems unnecessary considering the stock client and server only ever send movement packets from the client as unreliable. All other packets sent by the client and server are reliable and ordered.

As an aside, I doubt all packets currently sent as reliable actually require they are processed in the order they are sent? Perhaps a reliable unordered packet type would be useful.

## Why not discard all unreliable packets other than movement updates?

With this PR merged the client and server will discard most messages they receive via unreliable packets but not all. This is just to leave some space for future changes in the network protocol. It seems possible in the future some of those packets might be sent as unreliable. Since only one packet type is currently sent as unreliable, these restrictions can be increased in the future, but they can't as easily be loosened again. 

# To do

- More unit tests?
- Testing
- Measure pref impact on a full server? Should be negligible.

# How to test

- Build  the client and verify you can use the client without any issue:
  1. with servers without this PR 
  2. with servers without this PR while making a new account.
  3. servers with this PR.
  
- Build the server and check you can connect using:
   1. Standard unpatched clients
   2. New clients while registering a new account.
   3. New clients while logging into an existing account.
   
- Capture traffic for analysis and verify it appears to be indistinguishable from random noise.
- Compile the client or server in debug mode and set `secure.dump_network_encryption_key` to enabled, verify you can now decrypt traffic between one encrypted peer and the server if you follow the same steps as outlined in `network/encryption.h`.
A modified wireshark plugin is provided but requires you compile a build of https://github.com/Lekensteyn/luagcrypt for your wireshark installation to work.
- Try to modified a packet and verify the recipient rejects it.
- Send an unencrypted packet to a host expecting an encrypted one and verify it is rejected.
- Try to send a chat message or similar via an unreliable packet and verify the peer rejects it.
